### PR TITLE
Throw typed conditions for out-of-bounds indexing

### PR DESCRIPTION
**Pull request by lionel-**
Out-of-bounds indexing now throws errors of classes:

- `vctrs_error_index_oob_position` for OOB positions.
- `vctrs_error_index_oob_name` for OOB names.

Also tweaks [message wordings](https://github.com/r-lib/vctrs/compare/master...lionel-:add-oob-error-class?expand=1#diff-05f2962e6494e58a5cc1beaa5b135846).

### CODE REVIEW

File: R/conditions.R [hunk 1]
```diff
@@ -418,6 +418,65 @@ stop_names_must_be_unique <- function(locations) {
...
+#'   `vctrs_error_index_oob_name` containing fields `i` and `names`.
+#'
+#' @param i For `stop_index_oob_position()`, a number. For
+#'   `stop_index_oob_name()`, a name.
+#' @param size The length of the vector to subset from.
+#' @inheritParams rlang::abort
+#' @export
+stop_index_oob_position <- function(i, size, ..., .subclass = NULL) {
+  abort(
+    "",
```
Comment (hadley) [inline R/conditions.R]: Should we make this arg optional in `abort()`?

File: R/conditions.R [hunk 2]
```diff
@@ -418,6 +418,65 @@ stop_names_must_be_unique <- function(locations) {
...
+    )
+  }
+  NextMethod()
+}
+#' @export
+conditionMessage.vctrs_error_index_oob_name <- function(c) {
+  if (!nzchar(c$message)) {
+    c$message <- glue_lines(
+      "Must index an existing element.",
+      "* Can't subset unknown element `{c$i}`."
```
Comment (hadley) [inline R/conditions.R]:
```suggestion
      "* Can't subset element with unknown name `{c$i}`."
```

File: R/conditions.R (hunk 1 above)
Comment (lionel-) [reply R/conditions.R]: I think so as well, I opened an issue: https://github.com/r-lib/rlang/issues/840

**Comment by lionel-** [comment]
@hadley with the last commit, the stoppers are now vectorised:

```r
vctrs::vec_slice(letters, c(100, 1000))
#> Error: Must index existing elements.
#> * There are only 26 elements.
#> * Can't subset positions 100 and 1000

vctrs::vec_slice(letters, c(1, 100:103, 2, 104:110))
#> Error: Must index existing elements.
#> * There are only 26 elements.
#> * Can't subset positions 100, 101, 102, 103, 104, ...

vctrs::vec_slice(set_names(letters), c("foo", "bar"))
#> Error: Must index existing elements.
#> * Can't subset elements with unknown names `foo` and `bar`.

vctrs::vec_slice(set_names(letters), toupper(letters))
#> Error: Must index existing elements.
#> * Can't subset elements with unknown names `A`, `B`, `C`, `D`, `E`, ...
```

**Pull request merged**